dashboard/app/assets/javascripts/callouts.js

```diff
@@ -1,12 +1,53 @@
-dashboard.createCallouts = function(callouts) {
-  if (!callouts) {
-    return;
+(function () {
+
+  /**
+   * Snap all callouts to their target positions.  Keeps them in
+   * position when blockspace is scrolled.
+   */
+  function snapCalloutsToTargets() {
+    var triggerEvent = null;
+    var animate = false;
+    $('.cdo-qtips').qtip('reposition', triggerEvent, animate);
   }
 
-  // Hide callouts when the function editor is closed (otherwise they jump to the top left corner)
-  $(window).on('function_editor_closed', function() {
-    $('.cdo-qtips').qtip('hide');
-  });
+  /**
+   * For callouts with targets in the codeWorkspace (blockly, flyout elements,
+   * function editor elements, etc) hides callouts with targets that are
+   * scrolled out of view, and shows them again when they are scrolled back in
+   * to view.
+   * @function
+   */
+  var showOrHideCalloutsByTargetVisibility = (function () {
+    // Close around this object, which we use to remember which callouts
+    // were hidden by scrolling and should be shown again when they scroll
+    // back in.
+    /** @type {Object.<string, boolean>} */
+    var calloutsHiddenByScrolling = {};
+    return function () {
+      var codeWorkspace = $('#codeWorkspace');
```

Comment:
It might be an improvement to this particular method, but global dependence on `#codeWorkspace` is already a bit of a mess.  In this case, the only caller is [later in this file](https://github.com/code-dot-org/code-dot-org/pull/3117/files#diff-5ddf20d04374fa7973d9085c514402e3R78), a response to an event fired on document which doesn't have any more claim to `#codeWorkspace` than this method does.

The div in question is defined as part of the [standard page template](https://github.com/code-dot-org/code-dot-org/blob/cc9f22b0b95b6cf0d0a8b5dfe4e729af2c2f2bbc/apps/src/templates/page.html.ejs#L83) loaded by all of our apps.  Its existence is already a baked-in assumption [in StudioApp.js](https://github.com/code-dot-org/code-dot-org/blob/cc9f22b0b95b6cf0d0a8b5dfe4e729af2c2f2bbc/apps/src/StudioApp.js#L815).